### PR TITLE
PR for desktop app

### PR DESCRIPTION
**Pull request by amitpanwar789**

### CODE REVIEW

File: frontend/src/services/Native/index.d.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ declare global {
     ReactNativeWebView?: ReactNativeWebView;
     NativeRobosats?: NativeRobosats;
     RobosatsSettings: 'web-basic' | 'web-pro' | 'selfhosted-basic' | 'selfhosted-pro';
+    DesktopRobosats: undefined | 'Desktop-App';
```
Comment (KoalaSat) [inline frontend/src/services/Native/index.d.ts:8]: maybe we can generalize this do it more similar to `RobosatsSettings`, if we call this `RobosatsClient: 'desktop-app'` will be more readable and I can do the same for Mobile in another PR, what do you think?

File: frontend/src/services/System/SystemDesktopClient/index.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      // navigator clipboard api method'
+      void navigator.clipboard.writeText(value);
+    } else {
+      // text area method
+      const textArea = document.createElement('textarea');
+      textArea.value = value;
+      // make the textarea out of viewport
+      textArea.style.position = 'fixed';
+      textArea.style.left = '-999999px';
+      textArea.style.top = '-999999px';
```
Comment (KoalaSat) [inline frontend/src/services/System/SystemDesktopClient/index.ts:20]: Maybe you can achieve the same with just `visibility = hidden;`?

File: frontend/src/services/System/index.ts [hunk 3]
```diff
@@ -17,4 +18,4 @@ export const systemClient: SystemClient =
   // react-native-web view of the RoboSats Android app.
   window.navigator.userAgent.includes('robosats')
     ? new SystemNativeClient()
-    : new SystemWebClient();
+    : window.navigator.userAgent.includes('Electron')? new SystemDesktopClient() : new SystemWebClient();
```
Comment (KoalaSat) [inline frontend/src/services/System/index.ts:21]: Same about ternaries

**Review by KoalaSat — COMMENTED** [review]
Awesome job 🎉 ! I only reviewed the `frontend` folder. I'll take a look later to `desktopApp`

File: frontend/src/components/RobotAvatar/index.tsx [hunk 4]
```diff
@@ -71,7 +71,7 @@ const RobotAvatar: React.FC<Props> = ({
 
   useEffect(() => {
     if (shortAlias !== undefined) {
-      if (window.NativeRobosats === undefined) {
+      if (window.NativeRobosats === undefined) {
```
[the remark this answers is not in the document]
Comment (amitpanwar789) [reply frontend/src/components/RobotAvatar/index.tsx:74]: ohh i forget let me fix it

File: frontend/src/services/Native/index.d.ts (hunk 1 above)
Comment (amitpanwar789) [reply frontend/src/services/Native/index.d.ts:8]: i didn't understand

Comment (KoalaSat) [reply frontend/src/services/Native/index.d.ts:8]: Change this línea for:
```
RobosatsClient: 'desktop-app'
```
And where you are checking in other parts for `DesktopRobosats` switch it to `RobosatsClient === 'desktop-app'` that way it's more readable and will help other platforms such as the mobile android to reuse it

Comment (amitpanwar789) [reply frontend/src/services/Native/index.d.ts:8]: ok i guess then i need to change in html file also, to window.RobosatsClient = 'desktop-app'

Comment (amitpanwar789) [reply frontend/src/services/Native/index.d.ts:8]: and replace every use of window.DesktopRobosats === 'Desktop-App' with RobosatsClient === 'desktop-app'

Comment (KoalaSat) [reply frontend/src/services/Native/index.d.ts:8]: That's correct

**Comment by amitpanwar789** [comment]
@KoalaSat  have a look now .

File: desktopApp/app.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
+// Modules to control application life and create native browser window
```
Comment (KoalaSat) [inline desktopApp/app.js:1]: Did you tried to change this file to Typescript?

File: desktopApp/index.js.map [hunk 6]
```diff
@@ -0,0 +1 @@
...AoD;IACpD,UAAU,CAAC,WAAW,CAAC,EAAE,CAAC,eAAe,EAAE;QACzC,OAAO,CAAC,GAAG,CAAC,sCAAsC,CAAC,CAAC;QACpD,UAAU,CAAC,OAAO,CAAC,iBAAU,SAAS,uBAAoB,CAAC,CAAC;IAC9D,CAAC,CAAC,CAAC;IAEH,oDAAoD;IACpD,yCAAyC;AAC3C,CAAC;AAED,kEAAkE;AAClE,cAAG,CAAC,SAAS,EAAE,CAAC,IAAI,CAAC;IACnB,2CAA2C;IAC3C,YAAY,EAAE,CAAC;IAEf,2GAA2G;IAC3G,cAAG,CAAC,EAAE,CAAC,UAAU,EAAE;QACjB,IAAI,wBAAa,CAAC,aAAa,EAAE,CAAC,MAAM,KAAK,CAAC;YAAE,YAAY,EAAE,CAAC;IACjE,CAAC,CAAC,CAAC;AACL,CAAC,CAAC,CAAC;AAEH,+CAA+C;AAC/C,cAAG,CAAC,EAAE,CAAC,OAAO,EAAE;IACd,oCAAoC;IACpC,kBAAO,CAAC,cAAc,CAAC,UAAU,CAAC,eAAe,CAAC,EAAE,IAAI,EAAE,CAAC,kBAAkB,CAAC,EAAE,EAAE,UAAC,OAAO,EAAE,QAAQ;QAClG,IAAM,GAAG,GAAG,OAAO,CAAC,GAAG,CAAC;QACxB,IAAM,WAAW,GAAG,GAAG,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC;QACjC,IAAM,cAAc,GAAG,IAAI,CAAC,IAAI,CAAC,SAAS,EAAE,WAAW,CAAC,CAAC;QACzD,QAAQ,CAAC,EAAE,WAAW,EAAE,iBAAU,cAAc,CAAE,EAAE,CAAC,CAAC;IACxD,CAAC,CAAC,CAAC;IAEH,qDAAqD;IACrD,kBAAO,CAAC,cAAc,CAAC,QAAQ,CAAC;QAC9B,UAAU,EAAE,wBAAwB;QACpC,gBAAgB,EAAE,SAAS;KAC5B,CAAC,CAAC;AACL,CAAC,CAAC,CAAC;AAEH,kDAAkD;AAClD,cAAG,CAAC,EAAE,CAAC,mBAAmB,EAAE;IAC1B,yCAAyC;IACzC,GAAG,aAAH,GAAG,uBAAH,GAAG,CAAE,IAAI,EAAE,CAAC;IACZ,IAAI,OAAO,CAAC,QAAQ,KAAK,QAAQ;QAAE,cAAG,CAAC,IAAI,EAAE,CAAC;AAChD,CAAC,CAAC,CAAC"}
```
Comment (KoalaSat) [inline desktopApp/index.js.map:1]: Is this necesary?

**Review by KoalaSat — COMMENTED** [review]
Awesome job with /desktopApp I just have one quick question, I see a lot of duplications in `tor/tor-*`: 
- Are those README files necessary/different?
- Are the files under `/data`diferents per architecture? if not, maybe you can just move it up and use a symlink

File: frontend/src/basic/RobotPage/RobotProfile.tsx [hunk 7]
```diff
@@ -208,9 +208,9 @@ const RobotProfile = ({
           </Grid>
         ) : null}
 
-        {!Boolean(robot?.activeOrderId) &&
+        {!robot?.activeOrderId &&
         slot?.hashId &&
-        !Boolean(robot?.lastOrderId) &&
+        !robot?.lastOrderId &&
```
Comment (KoalaSat) [inline frontend/src/basic/RobotPage/RobotProfile.tsx:213]: Any reason for these changes?

File: frontend/src/components/RobotAvatar/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (KoalaSat) [reply frontend/src/components/RobotAvatar/index.tsx:74]: Still present

File: frontend/src/basic/RobotPage/RobotProfile.tsx (hunk 7 above)
Comment (amitpanwar789) [reply frontend/src/basic/RobotPage/RobotProfile.tsx:213]: This happens because I did lint:fix

File: desktopApp/app.js (hunk 5 above)
Comment (amitpanwar789) [reply desktopApp/app.js:1]: just ignore this file i am gonna delete this i have converted it in ts as index.ts

File: desktopApp/index.js.map (hunk 6 above)
Comment (amitpanwar789) [reply desktopApp/index.js.map:1]: when i compile the ts file  i got this file

**Comment by amitpanwar789** [comment]
https://github.com/RoboSats/robosats/pull/1299#discussion_r1640201233           i thought i only need to change for window.desktopApp ?

**Comment by KoalaSat** [comment]
> [#1299 (comment)](https://github.com/RoboSats/robosats/pull/1299#discussion_r1640201233) i thought i only need to change for window.desktopApp ?

I wad talking ahout the blank space, if you ran `lint:fix`, al good

**Comment by amitpanwar789** [comment]
> > [#1299 (comment)](https://github.com/RoboSats/robosats/pull/1299#discussion_r1640201233) i thought i only need to change for window.desktopApp ?
> 
> I wad talking ahout the blank space, if you ran `lint:fix`, al good

I already run that don't know why it is there

**Comment by amitpanwar789** [comment]
> Awesome job with `/desktopApp`! I just have one quick question, I see a lot of apparent duplication in `tor/tor-*`:
> 
> * Are those README files necessary/different?
> * Are the files under `/data`diferent per architecture? if not, maybe you can just move it up and use a symlink

which readme file

**Comment by KoalaSat** [comment]
> > Awesome job with `/desktopApp`! I just have one quick question, I see a lot of apparent duplication in `tor/tor-*`:
> > 
> > * Are those README files necessary/different?
> > * Are the files under `/data`diferent per architecture? if not, maybe you can just move it up and use a symlink
> 
> which readme file

For example `desktopApp/tor/tor-linux/tor/pluggable_transports/README.CONJURE.md`

**Comment by amitpanwar789** [comment]
> > > Awesome job with `/desktopApp`! I just have one quick question, I see a lot of apparent duplication in `tor/tor-*`:
> > > 
> > > * Are those README files necessary/different?
> > > * Are the files under `/data`diferent per architecture? if not, maybe you can just move it up and use a symlink
> > 
> > 
> > which readme file
> 
> For example `desktopApp/tor/tor-linux/tor/pluggable_transports/README.CONJURE.md`

Yes Readme are same in all three tor/

**Comment by KoalaSat** [comment]
> > > > Awesome job with `/desktopApp`! I just have one quick question, I see a lot of apparent duplication in `tor/tor-*`:
> > > > 
> > > > * Are those README files necessary/different?
> > > > * Are the files under `/data`diferent per architecture? if not, maybe you can just move it up and use a symlink
> > > 
> > > 
> > > which readme file
> > 
> > 
> > For example `desktopApp/tor/tor-linux/tor/pluggable_transports/README.CONJURE.md`
> 
> Yes Readme are same in all three tor/

Then maybe we can move it up to the parent folder and avoid duplicated files

**Comment by KoalaSat** [comment]
All good now good job with the duplications, I wanted to test it but I realized I have no idea how to run it 😄 can you add a `README.md` file to `/desktopApp` with a some breve instructions?

**Comment by amitpanwar789** [comment]
> All good now good job with the duplications, I wanted to test it but I realized I have no idea how to run it 😄 can you add a `README.md` file to `/desktopApp` with a some breve instructions?

When I try to test it in windows and clone it in windows but the symlink is not working

**Comment by KoalaSat** [comment]
@amitpanwar789 Tested and it looks great!! There is only one missing part on my opinion, did you configured the icon?
image

**Comment by amitpanwar789** [comment]
> @amitpanwar789 Tested on OSx and it looks great!! There is only one missing part on my opinion, did you configured the icon? image

for icon we need to make the icon type .ico for window and .icns for mac to configure it correctly

**Comment by KoalaSat** [comment]
> > @amitpanwar789 Tested on OSx and it looks great!! There is only one missing part on my opinion, did you configured the icon? image
> 
> for icon we need to make the icon type .ico for window and .icns for mac to configure it correctly

You can create then with this one https://github.com/RoboSats/robosats/blob/main/nodeapp/assets/icon/Robosats.svg

**Comment by amitpanwar789** [comment]
> > > @amitpanwar789 Tested on OSx and it looks great!! There is only one missing part on my opinion, did you configured the icon? image
> > 
> > 
> > for icon we need to make the icon type .ico for window and .icns for mac to configure it correctly
> 
> You can create then with this one https://github.com/RoboSats/robosats/blob/main/nodeapp/assets/icon/Robosats.svg

I have test

File: desktopApp/Readme.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+```
+
+### Step 4: Package the App
+
+To package the app for different platforms (Linux, Windows, macOS), use the corresponding npm commands:
+
+```bash
+npm run package:linux
+npm run package:win
+npm run package:mac
```
Comment (KoalaSat) [inline desktopApp/Readme.md:36]: `npm run package-mac` and I guess same for tohers

**Review by KoalaSat — APPROVED** [review]
Build tested in mac and it works! 🚀

**Comment by amitpanwar789** [comment]
should i make the separate pr for build part of add it in this pr

File: desktopApp/yarn.lock [hunk 9]
```diff
@@ -0,0 +1,2641 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (KoalaSat) [inline desktopApp/yarn.lock:1]: I just realized this, we don't use yarn on the project, I like yarn, but just to be consistent, can you use `npm` instead?

**Review by KoalaSat — APPROVED** [review]

**Comment by amitpanwar789** [comment]
@KoalaSat  i have a doubt and making build and uploading do we have to pre-release it as in android and then change release.yaml file download it as with android build

**Comment by KoalaSat** [comment]
> @KoalaSat i have a doubt and making build and uploading do we have to pre-release it as in android and then change release.yaml file download it as with android build

Desktop and Android app changes will be really in sync, so whatever process we are following in Android we should also folow it with Desktop :+1:

**Comment by amitpanwar789** [comment]
@KoalaSat , i have added the configuration for release please have a look , i haven't tested it locally so i am not sure

File: .github/workflows/desktop-build.yml [hunk 10]
```diff
@@ -2,6 +2,19 @@ name: "Build: Desktop"
...
+        type: string
+    secrets:
+      KEYSTORE:
+        required: true
+      KEY_ALIAS:
+        required: true
+      KEY_PASS:
+        required: true
+      KEY_STORE_PASS:
+        required: true
```
[the remark this answers is not in the document]
Comment (amitpanwar789) [reply .github/workflows/desktop-build.yml:17]: apart for this, all other things is good ? @KoalaSat

**Comment by KoalaSat** [comment]
Hey @amitpanwar789 thanks for the great effort! tested and approved :rocket:

**Pull request merged**